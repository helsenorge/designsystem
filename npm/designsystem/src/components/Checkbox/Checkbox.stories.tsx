import React from 'react';

import { ComponentStory, ComponentMeta } from '@storybook/react';

import { mapToBackgoundColor } from '../../../.storybook/StoryBackground';
import { FormMode, FormVariant } from '../../constants';
import Checkbox from './Checkbox';

export default {
  title: 'Components/Checkbox',
  component: Checkbox,
  argTypes: {
    label: {
      control: 'text',
      defaultValue: 'Jeg er frisk!',
    },
    checked: {
      control: 'boolean',
      defaultValue: false,
    },
    disabled: {
      control: 'boolean',
      defaultValue: false,
    },
    mode: {
      control: 'select',
      options: FormMode,
      defaultValue: FormMode.onwhite,
    },
    variant: {
      control: 'select',
      options: FormVariant,
      defaultValue: FormVariant.normal,
    },
    name: {
      control: 'text',
      defaultValue: '',
    },
    value: {
      control: 'text',
      defaultValue: '',
    },
    required: {
      control: 'boolean',
      defaultValue: false,
    },
  },
} as ComponentMeta<typeof Checkbox>;

export const Default: ComponentStory<typeof Checkbox> = (args: any) => (
  <div
    style={{
      width: '40rem',
    }}
  >
    <Checkbox {...args} />
  </div>
);
