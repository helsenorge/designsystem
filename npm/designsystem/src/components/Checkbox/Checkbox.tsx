import React, { useEffect, useState } from 'react';

import classNames from 'classnames';

import { AnalyticsId, FormMode, FormVariant, IconSize } from '../../constants';
import { usePseudoClasses } from '../../hooks/usePseudoClasses';
import { getColor } from '../../theme/currys/color';
import { uuid } from '../../utils/uuid';
import Icon from '../Icons';
import Check from '../Icons/Check';
import { getLabelText, renderLabelAsParent } from '../Label';

import checkboxStyles from './styles.module.scss';

export interface CheckboxProps
  extends Pick<
    React.InputHTMLAttributes<HTMLInputElement>,
    'aria-describedby' | 'name' | 'value' | 'disabled' | 'checked' | 'required' | 'onChange'
  > {
  /** Adds custom classes to the element. */
  className?: string;
  /** The <Label/> next to the checkbox - sublabels kan ikke kombineres med bigform variant */
  label: React.ReactNode;
  /** input id of the checkbox */
  inputId?: string;
  /** Changes the visuals of the checkbox */
  mode?: keyof typeof FormMode;
  /** Changes the visuals of the checkbox */
  variant?: keyof typeof FormVariant;
  /** Activates Error style for the checkbox - This is can be true while errorText is empty, when in a FormGroup */
  error?: boolean;
  /** Error text to show above the component */
  errorText?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
}

export const Checkbox = React.forwardRef((props: CheckboxProps, ref: React.Ref<HTMLInputElement>) => {
  const {
    className,
    checked = false,
    disabled,
    label,
    inputId = uuid(),
    mode = FormMode.onwhite,
    name = inputId,
    variant,
    errorText,
    error = !!errorText,
    value = getLabelText(label),
    testId,
    required,
    onChange,
    ...rest
  } = props;
  const [isChecked, setIsChecked] = useState(checked);
  const onWhite = mode === FormMode.onwhite;
  const onGrey = mode === FormMode.ongrey;
  const onBlueberry = mode === FormMode.onblueberry;
  const onInvalid = error || mode === FormMode.oninvalid;
  const onDark = mode === FormMode.ondark;
  const bigform = variant === FormVariant.bigform;
  const { refObject, isFocused } = usePseudoClasses(ref as React.RefObject<HTMLInputElement>);
  const checkboxWrapperClasses = classNames(checkboxStyles['checkbox-wrapper'], {
    [checkboxStyles['checkbox-wrapper--with-error']]: errorText,
    [checkboxStyles['checkbox-wrapper--bigform']]: bigform,
  });
  const checkboxLabelClasses = classNames(checkboxStyles['checkbox-label'], {
    [checkboxStyles['checkbox-label--disabled']]: disabled,
    [checkboxStyles['checkbox-label--on-dark']]: onDark,
    [checkboxStyles['checkbox-label--bigform']]: bigform,
    [checkboxStyles['checkbox-label__big-form--checked']]: bigform && isChecked,
    [checkboxStyles['checkbox-label__big-form--focus']]: bigform && isFocused,
    [checkboxStyles['checkbox-label__big-form--on-white']]: bigform && onWhite,
    [checkboxStyles['checkbox-label__big-form--on-grey']]: bigform && onGrey,
    [checkboxStyles['checkbox-label__big-form--on-blueberry']]: bigform && onBlueberry,
    [checkboxStyles['checkbox-label__big-form--on-invalid']]: bigform && onInvalid,
    [checkboxStyles['checkbox-label__big-form--disabled']]: bigform && disabled,
  });
  const checkboxClasses = classNames(checkboxStyles.checkbox, className);
  const checkboxIconWrapperClasses = classNames(checkboxStyles['checkbox__icon-wrapper'], {
    [checkboxStyles['checkbox__icon-wrapper--on-white']]: onWhite,
    [checkboxStyles['checkbox__icon-wrapper--on-grey']]: onGrey,
    [checkboxStyles['checkbox__icon-wrapper--on-invalid']]: onInvalid,
    [checkboxStyles['checkbox__icon-wrapper--disabled']]: disabled,
    [checkboxStyles['checkbox__icon-wrapper__regular--checked']]: !bigform && isChecked,
    [checkboxStyles['checkbox__icon-wrapper__regular--invalid']]: !bigform && isChecked && onInvalid,
    [checkboxStyles['checkbox__icon-wrapper__regular--on-dark']]: !bigform && isChecked && onDark,
    [checkboxStyles['checkbox__icon-wrapper__big-form--checked']]: bigform && isChecked,
    [checkboxStyles['checkbox__icon-wrapper__big-form--invalid']]: bigform && onInvalid,
    [checkboxStyles['checkbox__icon-wrapper--on-dark']]: onDark,
    [checkboxStyles['checkbox__icon-wrapper--on-blueberry']]: onBlueberry,
    [checkboxStyles['checkbox__icon-wrapper--invalid']]: onInvalid,
    [checkboxStyles['checkbox__icon-wrapper__big-form--invalid']]: bigform && isChecked && onInvalid,
    [checkboxStyles['checkbox__icon-wrapper__big-form--disabled']]: disabled && bigform && isChecked,
    [checkboxStyles['checkbox__icon-wrapper__big-form--checked--invalid']]: bigform && isChecked && onInvalid,
    [checkboxStyles['checkbox__icon-wrapper__big-form--checked--disabled']]: disabled && bigform && isChecked,
  });
  const labelTextClasses = classNames(checkboxStyles['checkbox-label__text'], {
    [checkboxStyles['checkbox-label__text__big-form--checked']]: bigform && isChecked,
    [checkboxStyles['checkbox-label__text__big-form--invalid']]: bigform && isChecked && onInvalid,
    [checkboxStyles['checkbox-label__text--on-dark']]: onDark,
    [checkboxStyles['checkbox-label__text--disabled']]: disabled,
  });
  const errorStyles = classNames(checkboxStyles['checkbox-errors']);

  let iconColor = getColor('white');
  if (onDark || (bigform && isChecked)) iconColor = getColor('blueberry', 900);
  if (onInvalid && bigform && isChecked) iconColor = getColor('white');
  if (disabled) iconColor = getColor('neutral', 400);

  useEffect(() => {
    setIsChecked(checked);
  }, [checked]);

  const onChangeHandler = (e: React.ChangeEvent<HTMLInputElement>): void => {
    if (onChange) {
      onChange(e);
    }

    setIsChecked(!isChecked);
  };

  const getLabelContent = (): React.ReactNode => {
    return (
      <>
        <input
          id={inputId}
          name={name}
          className={checkboxClasses}
          type="checkbox"
          checked={isChecked}
          disabled={disabled}
          value={value}
          ref={refObject}
          aria-describedby={props['aria-describedby'] ?? undefined}
          aria-invalid={error}
          required={required}
          onChange={onChangeHandler}
          {...rest}
        />
        <span className={checkboxIconWrapperClasses}>
          {isChecked && <Icon color={iconColor} className={checkboxStyles['checkbox__icon']} svgIcon={Check} size={IconSize.XSmall} />}
        </span>
      </>
    );
  };

  return (
    <div data-testid={testId} data-analyticsid={AnalyticsId.Checkbox} className={checkboxWrapperClasses}>
      {errorText && <p className={errorStyles}>{errorText}</p>}
      {renderLabelAsParent(
        label,
        getLabelContent(),
        inputId,
        mode as FormMode,
        disabled,
        checkboxLabelClasses,
        labelTextClasses,
        checkboxStyles['checkbox-sublabel-wrapper'],
        bigform
      )}
    </div>
  );
});

Checkbox.displayName = 'Checkbox';

export default Checkbox;
