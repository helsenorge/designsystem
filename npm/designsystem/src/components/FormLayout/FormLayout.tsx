import React from 'react';

import classNames from 'classnames';

import { AnalyticsId, FormVariant } from '../../constants';

import formGroupStyles from './styles.module.scss';

export enum FormLayoutColumns {
  one = 'one',
  two = 'two',
  three = 'three',
  four = 'four',
  five = 'five',
}

export interface FormLayoutProps {
  /** The max number of columns that will exist if space is available */
  maxColumns?: keyof typeof FormLayoutColumns;
  /** Sets the minimum width for the columns in pixels - this determines how many will fit on a row */
  colMinWidth?: number;
  /** Items in the FormLayout component */
  children?: React.ReactNode;
  /** Changes the visuals of the formgroup */
  variant?: keyof typeof FormVariant;
  /** Adds custom classes to the element. */
  className?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
  /** Function that helps map the form children */
  mapHelper?: (child: React.ReactNode, index: number) => React.ReactNode;
}

export const FormLayout = React.forwardRef((props: FormLayoutProps, ref: React.ForwardedRef<HTMLElement>) => {
  const { maxColumns: columns = FormLayoutColumns.one, colMinWidth = 300, variant, className, mapHelper } = props;

  const cssVariable = { '--min-col-width': `${colMinWidth}px` } as React.CSSProperties;
  const formLayoutContainerClasses = classNames(
    formGroupStyles['form-layout-container'],
    {
      [formGroupStyles['form-layout-container--bigform']]: variant === FormVariant.bigform,
    },
    className
  );
  const formLayoutChildClasses = classNames(formGroupStyles['form-layout-child'], {
    [formGroupStyles['form-layout-child--two']]: columns === FormLayoutColumns.two,
    [formGroupStyles['form-layout-child--three']]: columns === FormLayoutColumns.three,
    [formGroupStyles['form-layout-child--four']]: columns === FormLayoutColumns.four,
    [formGroupStyles['form-layout-child--five']]: columns === FormLayoutColumns.five,
  });

  return (
    <div data-testid={props.testId} data-analyticsid={AnalyticsId.FormLayout} style={cssVariable} className={formLayoutContainerClasses}>
      {React.Children.map(props.children, (child: React.ReactNode, index: number) => {
        return <div className={formLayoutChildClasses}>{mapHelper ? mapHelper(child, index) : child}</div>;
      })}
    </div>
  );
});

FormLayout.displayName = 'FormLayout';

export default FormLayout;
