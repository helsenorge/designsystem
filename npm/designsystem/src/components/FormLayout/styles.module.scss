@import '../../scss/spacers';
@import '../../scss/breakpoints';

.form-layout-container {
  // Settes med harde verdier fordi calc() lenger nede ikke st√∏tter getSpacer()
  --column-gap: 2rem;

  display: flex;
  flex-wrap: wrap;
  gap: getSpacer(s) var(--column-gap);

  &--large {
    row-gap: getSpacer(2xs);
  }
}

.form-layout-child {
  min-width: var(--min-col-width);
  flex-basis: calc(100% - var(--column-gap));

  &--two {
    flex-basis: calc(50% - var(--column-gap));
  }

  &--three {
    flex-basis: calc(33% - var(--column-gap));
  }

  &--four {
    flex-basis: calc(25% - var(--column-gap));
  }

  &--five {
    flex-basis: calc(20% - var(--column-gap));
  }
}
