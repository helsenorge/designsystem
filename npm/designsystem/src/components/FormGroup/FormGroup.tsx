import React, { useState } from 'react';

import classNames from 'classnames';

import { AnalyticsId, FormMode, FormVariant } from '../../constants';
import { isComponent } from '../../utils/component';
import uuid from '../../utils/uuid';
import Checkbox, { CheckboxProps } from '../Checkbox/Checkbox';
import ErrorWrapper from '../ErrorWrapper';
import FormLayout, { FormLayoutProps } from '../FormLayout';
import Input, { InputProps } from '../Input/Input';
import RadioButton, { RadioButtonProps, getRadioLabelClasses } from '../RadioButton/RadioButton';
import Select, { SelectProps } from '../Select';
import Title from '../Title';

import formGroupStyles from './styles.module.scss';

export type FormGroupTags = 'fieldset' | 'div';

export interface FormGroupProps {
  /** title for the the fieldset */
  title?: string;
  /** text placed in the legend tag of the fieldset */
  legend?: string;
  /** Items in the FormGroup component */
  children?: React.ReactNode;
  /** Adds custom classes to the element. */
  className?: string;
  /** Adds custom classes to the errorWrapper. */
  errorWrapperClassName?: string;
  /** Adds custom classes to the fieldset element. */
  fieldsetClassName?: string;
  /** Changes the visuals of the formgroup */
  mode?: keyof typeof FormMode;
  /** Changes the visuals of the formgroup */
  variant?: keyof typeof FormVariant;
  /** Error message */
  error?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
  /** Sets the data-testid attribute for the error-wrapper. */
  errorWrapperTestId?: string;
  /** Unique name for the child input element */
  name?: string;
  /** Unique name for the fieldset */
  fieldsetName?: string;
  /** Sets div instead of fieldset tag */
  htmlMarkup?: FormGroupTags;
  /** Renders the error component (Default: true) */
  renderError?: boolean;
}

export const FormGroup = React.forwardRef((props: FormGroupProps, ref: React.ForwardedRef<HTMLDivElement>) => {
  const {
    className,
    fieldsetClassName,
    mode = FormMode.onwhite,
    variant = FormVariant.normal,
    error,
    name,
    htmlMarkup = 'fieldset',
    renderError = true,
    errorWrapperClassName,
    errorWrapperTestId,
  } = props;
  const [checkedRadioId, setCheckedRadioId] = useState<string>();
  const [radioGroupId] = useState<string>(uuid());
  const onDark = mode === FormMode.ondark;
  const bigform = variant === FormVariant.bigform;
  const formGroupWrapperClasses = classNames(
    formGroupStyles['form-group-wrapper'],
    {
      [formGroupStyles['form-group-wrapper--on-dark']]: onDark,
      [formGroupStyles['form-group-wrapper--invalid']]: error,
    },
    className
  );
  const titleClasses = classNames(formGroupStyles['form-group-wrapper__title'], {
    [formGroupStyles['form-group-wrapper__title--on-dark']]: onDark && !error,
    [formGroupStyles['form-group-wrapper__title--bigform']]: bigform,
  });
  const formGroupClasses = classNames(formGroupStyles['form-group']);

  const legendClasses = classNames(formGroupStyles['field-set__legend'], {
    [formGroupStyles['field-set__legend--on-dark']]: onDark && !error,
    [formGroupStyles['field-set__legend--bigform']]: bigform,
  });

  const fieldsetClasses = classNames(formGroupStyles['field-set'], fieldsetClassName);

  const mapFormComponent = (child: React.ReactNode, index: number): React.ReactNode => {
    if (isComponent<FormGroupProps>(child, FormLayout)) {
      return React.cloneElement(child as React.ReactElement<FormLayoutProps>, {
        variant,
        mapHelper: mapFormComponent,
      });
    } else if (isComponent<FormGroupProps>(child, FormGroup)) {
      return React.cloneElement(child, {
        mode,
        variant,
        error,
        renderError: false,
      });
    } else if (isComponent<CheckboxProps>(child, Checkbox)) {
      return React.cloneElement(child as React.ReactElement<CheckboxProps>, {
        name: name ?? child.props.name,
        mode,
        variant,
        error: !!error,
      });
    } else if (isComponent<RadioButtonProps>(child, RadioButton)) {
      const radioId = typeof child.props.inputId === 'undefined' ? radioGroupId + index : child.props.inputId;
      return React.cloneElement(child as React.ReactElement<RadioButtonProps>, {
        inputId: radioId,
        name: name ?? child.props.name,
        mode,
        onChange: (event: React.ChangeEvent<HTMLInputElement>) => {
          setCheckedRadioId(event.target.id);
          child.props.onChange && child.props.onChange(event);
        },
        variant,
        error: !!error,
        labelClassNames: getRadioLabelClasses(radioId, mode as FormMode, bigform, checkedRadioId),
      });
    } else if (isComponent<InputProps>(child, Input)) {
      return React.cloneElement(child as React.ReactElement<InputProps>, {
        name: name ?? child.props.name,
        mode,
        variant,
        error: !!error,
      });
    } else if (isComponent<SelectProps>(child, Select)) {
      return React.cloneElement(child as React.ReactElement<SelectProps>, {
        name: name ?? child.props.name,
        mode,
        error: !!error,
      });
    }
    return child;
  };

  const formGroupContent = (): React.ReactNode => {
    return (
      <div className={formGroupClasses}>
        {htmlMarkup === 'div' && (
          <div className={fieldsetClasses}>
            {props.legend && <h5 className={legendClasses}>{props.legend}</h5>}
            {React.Children.map(props.children, mapFormComponent)}
          </div>
        )}
        {htmlMarkup === 'fieldset' && (
          <fieldset name={props.fieldsetName} className={fieldsetClasses}>
            {props.legend && <legend className={legendClasses}>{props.legend}</legend>}
            {React.Children.map(props.children, mapFormComponent)}
          </fieldset>
        )}
      </div>
    );
  };

  return (
    <div data-testid={props.testId} data-analyticsid={AnalyticsId.FormGroup} className={formGroupWrapperClasses} ref={ref} tabIndex={-1}>
      {props.title && (
        <Title className={titleClasses} htmlMarkup={'h4'} appearance={'title4'} margin={{ marginTop: 0, marginBottom: error ? 1 : 2 }}>
          {props.title}
        </Title>
      )}
      {renderError ? (
        <ErrorWrapper className={errorWrapperClassName} errorText={error} testId={errorWrapperTestId}>
          {formGroupContent()}
        </ErrorWrapper>
      ) : (
        formGroupContent()
      )}
    </div>
  );
});

FormGroup.displayName = 'FormGroup';

export default FormGroup;
