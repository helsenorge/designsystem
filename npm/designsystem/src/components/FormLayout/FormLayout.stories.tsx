import React from 'react';

import { ComponentStory, ComponentMeta } from '@storybook/react';

import Checkbox from '../Checkbox/Checkbox';
import RadioButton from '../RadioButton/RadioButton';
import FormGroup from '../FormGroup';
import FormLayout, { FormLayoutColumns } from './FormLayout';
import GridExample from '../GridExample';

export default {
  title: 'Components/FormLayout',
  component: FormLayout,
  argTypes: {
    maxColumns: {
      control: 'select',
      options: FormLayoutColumns,
      defaultValue: FormLayoutColumns.three,
    },
    colMinWidth: {
      control: 'number',
      defaultValue: 300,
    },
  },
} as ComponentMeta<typeof FormLayout>;

export const CheckboxChildren: ComponentStory<typeof FormLayout> = (args: any) => (
  <GridExample>
    <FormGroup title={'Her kan du styre maks antall kolonner'} legend={'Antallet er basert på hvor mange som har plass'}>
      <FormLayout {...args}>
        <Checkbox inputId={'Checkbox1'} label={'Checkbox hei'} />
        <Checkbox inputId={'Checkbox2'} label={'Checkbox lalalala'} />
        <Checkbox inputId={'Checkbox3'} label={'Checkbox asdadasd afasasfaa'} />
        <Checkbox inputId={'Checkbox4'} label={'Checkbox hmm'} />
        <Checkbox inputId={'Checkbox5'} label={'Checkbox'} />
        <Checkbox inputId={'Checkbox6'} label={'Checkbox jadada'} />
      </FormLayout>
    </FormGroup>
  </GridExample>
);

export const RadioButtonChildren: ComponentStory<typeof FormGroup> = (args: any) => (
  <GridExample>
    <FormGroup
      title={'Her kan du styre maks antall kolonner'}
      legend={'Antallet er basert på hvor mange som har plass'}
      name={'radiogroup1'}
    >
      <FormLayout {...args}>
        <RadioButton inputId={'RadioButton1'} label={'RadioButton 1'} />
        <RadioButton inputId={'RadioButton2'} label={'RadioButton 2'} />
        <RadioButton inputId={'RadioButton3'} label={'RadioButton 3'} />
      </FormLayout>
    </FormGroup>
    <FormGroup legend={'Radio radio hello!'} name={'radiogroup2'}>
      <FormLayout {...args}>
        <RadioButton inputId={'RadioButton4'} label={'RadioButton 4'} />
        <RadioButton inputId={'RadioButton5'} label={'RadioButton 5'} />
        <RadioButton inputId={'RadioButton6'} label={'RadioButton 6'} />
      </FormLayout>
    </FormGroup>
  </GridExample>
);
