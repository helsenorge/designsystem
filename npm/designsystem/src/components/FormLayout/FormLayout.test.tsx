import React from 'react';

import { render, screen } from '@testing-library/react';

import FormLayout, { FormLayoutColumns } from './FormLayout';
import Checkbox from '../Checkbox/Checkbox';
import FormGroup from '../FormGroup';
import Label from '../Label';
import RadioButton from '../RadioButton';

describe('Gitt at FormLayout skal vises', (): void => {
  describe('Når FormLayout rendres', (): void => {
    test('Så vises FormLayout', (): void => {
      render(
        <FormGroup title={'One amazing title'} legend={'Check out these checkboxes!'}>
          <FormLayout>
            <Checkbox inputId={'Checkbox1'} label={<Label labelTexts={[{ text: 'Checkbox 1' }]} />} />
            <Checkbox inputId={'Checkbox2'} label={<Label labelTexts={[{ text: 'Checkbox 2' }]} />} />
            <Checkbox inputId={'Checkbox3'} label={<Label labelTexts={[{ text: 'Checkbox 3' }]} />} />
          </FormLayout>
        </FormGroup>
      );

      const layout =
        screen.getByText('Checkbox 1').parentElement?.parentElement?.parentElement?.parentElement?.parentElement?.parentElement
          ?.parentElement?.parentElement;
      expect(layout).toBeVisible();
      expect(layout?.className).toBe('form-layout-container');
    });
  });

  describe('Når maxColumns settes', (): void => {
    test('Så er stylingen satt riktig', (): void => {
      render(
        <FormGroup title={'One amazing title'} legend={'Check out these checkboxes!'}>
          <FormLayout maxColumns={FormLayoutColumns.two}>
            <Checkbox inputId={'Checkbox1'} label={<Label labelTexts={[{ text: 'Checkbox 1' }]} />} />
            <Checkbox inputId={'Checkbox2'} label={<Label labelTexts={[{ text: 'Checkbox 2' }]} />} />
            <Checkbox inputId={'Checkbox3'} label={<Label labelTexts={[{ text: 'Checkbox 3' }]} />} />
          </FormLayout>
        </FormGroup>
      );

      const layout =
        screen.getByText('Checkbox 1').parentElement?.parentElement?.parentElement?.parentElement?.parentElement?.parentElement
          ?.parentElement;
      expect(layout).toBeVisible();
      expect(layout?.className).toBe('form-layout-child form-layout-child--two');
    });
  });

  describe('Når colMinWidth settes', (): void => {
    test('Så er css variabel satt riktig', (): void => {
      render(
        <FormGroup title={'One amazing title'} legend={'Check out these checkboxes!'}>
          <FormLayout colMinWidth={200}>
            <Checkbox inputId={'Checkbox1'} label={<Label labelTexts={[{ text: 'Checkbox 1' }]} />} />
            <Checkbox inputId={'Checkbox2'} label={<Label labelTexts={[{ text: 'Checkbox 2' }]} />} />
            <Checkbox inputId={'Checkbox3'} label={<Label labelTexts={[{ text: 'Checkbox 3' }]} />} />
          </FormLayout>
        </FormGroup>
      );

      const layout =
        screen.getByText('Checkbox 1').parentElement?.parentElement?.parentElement?.parentElement?.parentElement?.parentElement
          ?.parentElement?.parentElement;
      expect(layout).toBeVisible();
      expect(layout?.style['_values']['--min-col-width']).toBe('200px');
    });
  });

  describe('Når variant settes', (): void => {
    test('Så er stylingen satt riktig på children', (): void => {
      render(
        <FormGroup title={'One amazing title'} legend={'Check out these checkboxes!'} variant={'bigform'}>
          <FormLayout maxColumns={FormLayoutColumns.two}>
            <Checkbox inputId={'Checkbox1'} label={<Label labelTexts={[{ text: 'Checkbox 1' }]} />} />
            <Checkbox inputId={'Checkbox2'} label={<Label labelTexts={[{ text: 'Checkbox 2' }]} />} />
            <Checkbox inputId={'Checkbox3'} label={<Label labelTexts={[{ text: 'Checkbox 3' }]} />} />
          </FormLayout>
        </FormGroup>
      );

      const checkbox = screen.getByText('Checkbox 1').parentElement?.parentElement?.parentElement;
      expect(checkbox).toBeVisible();
      expect(checkbox?.className).toBe('checkbox-label checkbox-label--bigform checkbox-label__big-form--on-white');
    });
  });

  describe('Når FormLayout har checkbox children', (): void => {
    test('Så rendres de', (): void => {
      render(
        <FormGroup title={'One amazing title'} legend={'Check out these checkboxes!'}>
          <FormLayout maxColumns={FormLayoutColumns.two}>
            <Checkbox inputId={'Checkbox1'} label={<Label labelTexts={[{ text: 'Checkbox 1' }]} />} />
            <Checkbox inputId={'Checkbox2'} label={<Label labelTexts={[{ text: 'Checkbox 2' }]} />} />
            <Checkbox inputId={'Checkbox3'} label={<Label labelTexts={[{ text: 'Checkbox 3' }]} />} />
          </FormLayout>
        </FormGroup>
      );

      const checkboxArray = screen.getAllByRole('checkbox');
      expect(checkboxArray.length).toBe(3);
    });
  });

  describe('Når FormLayout har radio children', (): void => {
    test('Så rendres de', (): void => {
      render(
        <FormGroup title={'One amazing title'} legend={'Check out these checkboxes!'}>
          <FormLayout maxColumns={FormLayoutColumns.two}>
            <RadioButton inputId={'Radio1'} label={<Label labelTexts={[{ text: 'Radio 1' }]} />} />
            <RadioButton inputId={'Radio2'} label={<Label labelTexts={[{ text: 'Radio 2' }]} />} />
            <RadioButton inputId={'Radio3'} label={<Label labelTexts={[{ text: 'Radio 3' }]} />} />
          </FormLayout>
        </FormGroup>
      );

      const radioButtonArray = screen.getAllByRole('radio');
      expect(radioButtonArray.length).toBe(3);
    });
  });
});
