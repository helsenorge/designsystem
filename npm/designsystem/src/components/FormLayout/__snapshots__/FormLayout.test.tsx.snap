// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at FormLayout skal vises Når FormLayout rendres Så vises FormLayout 1`] = `
<div>
  <div
    class="form-group-wrapper"
    data-analyticsid="form-group"
    tabindex="-1"
  >
    <h4
      class="title title--title4 form-group-wrapper__title"
      data-analyticsid="title"
      style="margin-top: 0rem; margin-bottom: 2rem;"
    >
      One amazing title
    </h4>
    <div
      class="error-wrapper"
    >
      <div
        class="form-group"
      >
        <fieldset
          class="field-set"
          name="One amazing title"
        >
          <legend
            class="field-set__legend"
          >
            Check out these checkboxes!
          </legend>
          <div
            class="form-layout-container"
            data-analyticsid="form-layout"
            style="--min-col-width: 300px;"
          >
            <div
              class="form-layout-child"
            >
              <div
                class="checkbox-wrapper"
                data-analyticsid="checkbox"
              >
                <label
                  class="checkbox-label"
                  for="Checkbox1"
                >
                  <input
                    aria-invalid="false"
                    class="checkbox"
                    id="Checkbox1"
                    name="valueOf"
                    type="checkbox"
                    value="Checkbox 1"
                  />
                  <span
                    class="checkbox__icon-wrapper"
                  />
                  <span
                    class="checkbox-label__text"
                  >
                    Checkbox 1
                  </span>
                </label>
              </div>
            </div>
            <div
              class="form-layout-child"
            >
              <div
                class="checkbox-wrapper"
                data-analyticsid="checkbox"
              >
                <label
                  class="checkbox-label"
                  for="Checkbox2"
                >
                  <input
                    aria-invalid="false"
                    class="checkbox"
                    id="Checkbox2"
                    name="valueOf"
                    type="checkbox"
                    value="Checkbox 2"
                  />
                  <span
                    class="checkbox__icon-wrapper"
                  />
                  <span
                    class="checkbox-label__text"
                  >
                    Checkbox 2
                  </span>
                </label>
              </div>
            </div>
            <div
              class="form-layout-child"
            >
              <div
                class="checkbox-wrapper"
                data-analyticsid="checkbox"
              >
                <label
                  class="checkbox-label"
                  for="Checkbox3"
                >
                  <input
                    aria-invalid="false"
                    class="checkbox"
                    id="Checkbox3"
                    name="valueOf"
                    type="checkbox"
                    value="Checkbox 3"
                  />
                  <span
                    class="checkbox__icon-wrapper"
                  />
                  <span
                    class="checkbox-label__text"
                  >
                    Checkbox 3
                  </span>
                </label>
              </div>
            </div>
          </div>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
